File: core/src/test/java/ch/dissem/bitmessage/entity/SerializationTest.java
/*
 * Copyright 2015 Christian Basler
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ch.dissem.bitmessage.entity;

import ch.dissem.bitmessage.entity.payload.*;
import ch.dissem.bitmessage.entity.valueobject.ExtendedEncoding;
import ch.dissem.bitmessage.entity.valueobject.InventoryVector;
import ch.dissem.bitmessage.entity.valueobject.Label;
import ch.dissem.bitmessage.factory.Factory;
import ch.dissem.bitmessage.utils.TestBase;
import ch.dissem.bitmessage.utils.TestUtils;
import org.junit.Test;

import java.io.*;
import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Collections;

import static ch.dissem.bitmessage.entity.Plaintext.Type.MSG;
import static ch.dissem.bitmessage.utils.Singleton.cryptography;
import static org.junit.Assert.*;

public class SerializationTest extends TestBase {
    @Test
    public void ensureGetPubkeyIsDeserializedAndSerializedCorrectly() throws IOException {
        doTest("V2GetPubkey.payload", 2, GetPubkey.class);
        doTest("V3GetPubkey.payload", 2, GetPubkey.class);
        doTest("V4GetPubkey.payload", 2, GetPubkey.class);
    }

    @Test
    public void ensureV2PubkeyIsDeserializedAndSerializedCorrectly() throws IOException {
        doTest("V2Pubkey.payload", 2, V2Pubkey.class);
    }

    @Test
    public void ensureV3PubkeyIsDeserializedAndSerializedCorrectly() throws IOException {
        doTest("V3Pubkey.payload", 3, V3Pubkey.class);
    }

    @Test
    public void ensureV4PubkeyIsDeserializedAndSerializedCorrectly() throws IOException {
        doTest("V4Pubkey.payload", 4, V4Pubkey.class);
    }

    @Test
    public void ensureV1MsgIsDeserializedAndSerializedCorrectly() throws IOException {
        doTest("V1Msg.payload", 1, Msg.class);
    }

    @Test
    public void ensureV4BroadcastIsDeserializedAndSerializedCorrectly() throws IOException {
        doTest("V4Broadcast.payload", 4, V4Broadcast.class);
    }

    @Test
    public void ensureV5BroadcastIsDeserializedAndSerializedCorrectly() throws IOException {
        doTest("V5Broadcast.payload", 5, V5Broadcast.class);
    }

    @Test
    public void ensureUnknownDataIsDeserializedAndSerializedCorrectly() throws IOException {
        doTest("V1MsgStrangeData.payload", 1, GenericPayload.class);
    }

    @Test
    public void ensurePlaintextIsSerializedAndDeserializedCorrectly() throws Exception {
        Plaintext p1 = new Plaintext.Builder(MSG)
            .from(TestUtils.loadIdentity("BM-2cSqjfJ8xK6UUn5Rw3RpdGQ9RsDkBhWnS8"))
            .to(TestUtils.loadContact())
            .message("Subject", "Message")
            .ackData("ackMessage".getBytes())
            .signature(new byte[0])
            .build();
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        p1.write(out);
        ByteArrayInputStream in = new ByteArrayInputStream(out.toByteArray());
        Plaintext p2 = Plaintext.read(MSG, in);

        // Received is automatically set on deserialization, so we'll need to set it to 0
        Field received = Plaintext.class.getDeclaredField("received");
        received.setAccessible(true);
        received.set(p2, 0L);

        assertEquals(p1, p2);
    }

    @Test
    public void ensurePlaintextWithExtendedEncodingIsSerializedAndDeserializedCorrectly() throws Exception {
        Plaintext p1 = new Plaintext.Builder(MSG)
            .from(TestUtils.loadIdentity("BM-2cSqjfJ8xK6UUn5Rw3RpdGQ9RsDkBhWnS8"))
            .to(TestUtils.loadContact())
            .message(new ExtendedEncoding.Builder().message()
                .subject("Subject")
                .body("Message")
                .build())
            .ackData("ackMessage".getBytes())
            .signature(new byte[0])
            .build();
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        p1.write(out);
        ByteArrayInputStream in = new ByteArrayInputStream(out.toByteArray());
        Plaintext p2 = Plaintext.read(MSG, in);

        // Received is automatically set on deserialization, so we'll need to set it to 0
        Field received = Plaintext.class.getDeclaredField("received");
        received.setAccessible(true);
        received.set(p2, 0L);

        assertEquals(p1, p2);
    }

    @Test
    public void ensurePlaintextWithAckMessageIsSerializedAndDeserializedCorrectly() throws Exception {
        Plaintext p1 = new Plaintext.Builder(MSG)
            .from(TestUtils.loadIdentity("BM-2cSqjfJ8xK6UUn5Rw3RpdGQ9RsDkBhWnS8"))
            .to(TestUtils.loadContact())
            .message("Subject", "Message")
            .ackData("ackMessage".getBytes())
            .signature(new byte[0])
            .build();
        ObjectMessage ackMessage1 = p1.getAckMessage();
        assertNotNull(ackMessage1);

        ByteArrayOutputStream out = new ByteArrayOutputStream();
        p1.write(out);
        ByteArrayInputStream in = new ByteArrayInputStream(out.toByteArray());
        Plaintext p2 = Plaintext.read(MSG, in);

        // Received is automatically set on deserialization, so we'll need to set it to 0
        Field received = Plaintext.class.getDeclaredField("received");
        received.setAccessible(true);
        received.set(p2, 0L);

        assertEquals(p1, p2);
        assertEquals(ackMessage1, p2.getAckMessage());
    }

    @Test
    public void ensureNetworkMessageIsSerializedAndDeserializedCorrectly() throws Exception {
        ArrayList<InventoryVector> ivs = new ArrayList<>(50000);
        for (int i = 0; i < 50000; i++) {
            ivs.add(new InventoryVector(cryptography().randomBytes(32)));
        }

        Inv inv = new Inv.Builder().inventory(ivs).build();
        NetworkMessage before = new NetworkMessage(inv);
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        before.write(out);

        NetworkMessage after = Factory.getNetworkMessage(3, new ByteArrayInputStream(out.toByteArray()));
        assertNotNull(after);
        Inv invAfter = (Inv) after.getPayload();
        assertEquals(ivs, invAfter.getInventory());
    }

    private void doTest(String resourceName, int version, Class<?> expectedPayloadType) throws IOException {
        byte[] data = TestUtils.getBytes(resourceName);
        InputStream in = new ByteArrayInputStream(data);
        ObjectMessage object = Factory.getObjectMessage(version, in, data.length);
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        assertNotNull(object);
        object.write(out);
        assertArrayEquals(data, out.toByteArray());
        assertEquals(expectedPayloadType.getCanonicalName(), object.getPayload().getClass().getCanonicalName());
    }

    @Test
    public void ensureSystemSerializationWorks() throws Exception {
        Plaintext plaintext = new Plaintext.Builder(MSG)
            .from(TestUtils.loadContact())
            .to(TestUtils.loadIdentity("BM-2cSqjfJ8xK6UUn5Rw3RpdGQ9RsDkBhWnS8"))
            .labels(Collections.singletonList(new Label("Test", Label.Type.INBOX, 0)))
            .message("Test", "Test Test.\nTest")
            .build();
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        ObjectOutputStream oos = new ObjectOutputStream(out);
        oos.writeObject(plaintext);

        ByteArrayInputStream in = new ByteArrayInputStream(out.toByteArray());
        ObjectInputStream ois = new ObjectInputStream(in);
        assertEquals(plaintext, ois.readObject());
    }
}

File: core/src/test/kotlin/ch/dissem/bitmessage/entity/SerializationTest.kt
/*
 * Copyright 2017 Christian Basler
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ch.dissem.bitmessage.entity

import ch.dissem.bitmessage.entity.Plaintext.Type.MSG
import ch.dissem.bitmessage.entity.payload.*
import ch.dissem.bitmessage.entity.valueobject.InventoryVector
import ch.dissem.bitmessage.entity.valueobject.Label
import ch.dissem.bitmessage.entity.valueobject.extended.Message
import ch.dissem.bitmessage.factory.Factory
import ch.dissem.bitmessage.utils.TestBase
import ch.dissem.bitmessage.utils.TestUtils
import org.hamcrest.Matchers.`is`
import org.junit.Assert.*
import org.junit.Test
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.util.*

class SerializationTest : TestBase() {
    @Test
    fun `ensure GetPubkey is deserialized and serialized correctly`() {
        doTest("V2GetPubkey.payload", 2, GetPubkey::class.java)
        doTest("V3GetPubkey.payload", 2, GetPubkey::class.java)
        doTest("V4GetPubkey.payload", 2, GetPubkey::class.java)
    }

    @Test
    fun `ensure V2Pubkey is deserialized and serialized correctly`() {
        doTest("V2Pubkey.payload", 2, V2Pubkey::class.java)
    }

    @Test
    fun `ensure V3Pubkey is deserialized and serialized correctly`() {
        doTest("V3Pubkey.payload", 3, V3Pubkey::class.java)
    }

    @Test
    fun `ensure V4Pubkey is deserialized and serialized correctly`() {
        doTest("V4Pubkey.payload", 4, V4Pubkey::class.java)
    }

    @Test
    fun `ensure V1 msg is deserialized and serialized correctly`() {
        doTest("V1Msg.payload", 1, Msg::class.java)
    }

    @Test
    fun `ensure V4Broadcast is deserialized and serialized correctly`() {
        doTest("V4Broadcast.payload", 4, V4Broadcast::class.java)
    }

    @Test
    fun `ensure V5Broadcast is deserialized and serialized correctly`() {
        doTest("V5Broadcast.payload", 5, V5Broadcast::class.java)
    }

    @Test
    fun `ensure unknown data is deserialized and serialized correctly`() {
        doTest("V1MsgStrangeData.payload", 1, GenericPayload::class.java)
    }

    @Test
    fun `ensure plaintext is serialized and deserialized correctly`() {
        val expected = Plaintext.Builder(MSG)
            .from(TestUtils.loadIdentity("BM-2cSqjfJ8xK6UUn5Rw3RpdGQ9RsDkBhWnS8"))
            .to(TestUtils.loadContact())
            .message("Subject", "Message")
            .ackData("ackMessage".toByteArray())
            .signature(ByteArray(0))
            .build()
        val out = ByteArrayOutputStream()
        expected.write(out)
        val `in` = ByteArrayInputStream(out.toByteArray())
        val actual = Plaintext.read(MSG, `in`)

        // Received is automatically set on deserialization, so we'll need to set it to null
        val received = Plaintext::class.java.getDeclaredField("received")
        received.isAccessible = true
        received.set(actual, null)

        assertThat(expected, `is`(actual))
    }

    @Test
    fun `ensure plaintext with extended encoding is serialized and deserialized correctly`() {
        val expected = Plaintext.Builder(MSG)
            .from(TestUtils.loadIdentity("BM-2cSqjfJ8xK6UUn5Rw3RpdGQ9RsDkBhWnS8"))
            .to(TestUtils.loadContact())
            .message(Message.Builder()
                .subject("Subject")
                .body("Message")
                .build())
            .ackData("ackMessage".toByteArray())
            .signature(ByteArray(0))
            .build()
        val out = ByteArrayOutputStream()
        expected.write(out)
        val `in` = ByteArrayInputStream(out.toByteArray())
        val actual = Plaintext.read(MSG, `in`)

        // Received is automatically set on deserialization, so we'll need to set it to null
        val received = Plaintext::class.java.getDeclaredField("received")
        received.isAccessible = true
        received.set(actual, null)

        assertEquals(expected, actual)
    }

    @Test
    fun `ensure plaintext with ack message is serialized and deserialized correctly`() {
        val expected = Plaintext.Builder(MSG)
            .from(TestUtils.loadIdentity("BM-2cSqjfJ8xK6UUn5Rw3RpdGQ9RsDkBhWnS8"))
            .to(TestUtils.loadContact())
            .message("Subject", "Message")
            .ackData("ackMessage".toByteArray())
            .signature(ByteArray(0))
            .build()
        val ackMessage1 = expected.ackMessage
        assertNotNull(ackMessage1)

        val out = ByteArrayOutputStream()
        expected.write(out)
        val `in` = ByteArrayInputStream(out.toByteArray())
        val actual = Plaintext.read(MSG, `in`)

        // Received is automatically set on deserialization, so we'll need to set it to null
        val received = Plaintext::class.java.getDeclaredField("received")
        received.isAccessible = true
        received.set(actual, null)

        assertEquals(expected, actual)
        assertEquals(ackMessage1, actual.ackMessage)
    }

    @Test
    fun `ensure network message is serialized and deserialized correctly`() {
        val ivs = ArrayList<InventoryVector>(50000)
        for (i in 0..49999) {
            ivs.add(TestUtils.randomInventoryVector())
        }

        val inv = Inv(ivs)
        val before = NetworkMessage(inv)
        val out = ByteArrayOutputStream()
        before.write(out)

        val after = Factory.getNetworkMessage(3, ByteArrayInputStream(out.toByteArray()))
        assertNotNull(after)
        val invAfter = after!!.payload as Inv
        assertEquals(ivs, invAfter.inventory)
    }

    private fun doTest(resourceName: String, version: Int, expectedPayloadType: Class<*>) {
        val data = TestUtils.getBytes(resourceName)
        val `in` = ByteArrayInputStream(data)
        val objectMessage = Factory.getObjectMessage(version, `in`, data.size)
        val out = ByteArrayOutputStream()
        assertNotNull(objectMessage)
        objectMessage!!.write(out)
        assertArrayEquals(data, out.toByteArray())
        assertEquals(expectedPayloadType.canonicalName, objectMessage.payload.javaClass.canonicalName)
    }

    @Test
    fun `ensure system serialization works`() {
        val plaintext = Plaintext.Builder(MSG)
            .from(TestUtils.loadContact())
            .to(TestUtils.loadIdentity("BM-2cSqjfJ8xK6UUn5Rw3RpdGQ9RsDkBhWnS8"))
            .labels(listOf(Label("Test", Label.Type.INBOX, 0)))
            .message("Test", "Test Test.\nTest")
            .build()
        val out = ByteArrayOutputStream()
        val oos = ObjectOutputStream(out)
        oos.writeObject(plaintext)

        val `in` = ByteArrayInputStream(out.toByteArray())
        val ois = ObjectInputStream(`in`)
        assertEquals(plaintext, ois.readObject())
    }
}
